cep/source/man/cepDate.3
.\" This manpage has been automatically generated by docbook2man 
.\" from a DocBook document.  This tool can be found at:
.\" <http://shell.ipoline.com/~elmert/comp/docbook2X/> 
.\" Please send any bug reports, improvements, comments, patches, 
.\" etc. to Steve Cheng <steve@ggi-project.org>.
.TH "cepDate" "3" "26 November 2002" "" ""
.SH NAME
cepDate \- Converts a decimal date in the format of YYYY.DDDD, where YYYY is a four
.SH SYNOPSIS

.nf
 To create a cepDate object to convert a decimal date to a Gregorian date use:-
 
 cepDate date(value);
 
 <para><itemizedlist>
 <listitem><para>value:- The decimal value to be converted. It must be of the form
 YYYY.DDDD where YYYY is the four digit year and DDDD is the decimal represenation
 of the number of days into the year</para></listitem>
 </itemizedlist></para>
 
 To create a cepDate object to convert a Gregorian date to a decimal date use:-
 
 cepDate date(day, month, year);
 
 <para><itemizedlist>
 <listitem><para>day:- interger value of the day of the month from 1..31</para></listitem>
 <listitem><para>month:- the name of the month. This can be either Jan...Dec or
 January...December</para></listitem>
 <listitem><para>year:- interger value of the year</para></listitem>
 </itemizedlist></para>
 
 </para>
 
 
.fi
.SH "DESCRIPTION"
.PP
An implementation of the <command>cepDate</command> class
converts a decimal date to a Gegorian date or a Gregorian date to a decimal.
</para>
<para>
<command>const string getDay()</command>
returns day of month in the format of 01...31. A return value of -1 indicates
an error because the date entered was invalid.
</para>
<para>
<command>const string getMonth()</command>
returns month in the format of 01...12. A return value of -1 indicates
an error because the date entered was invalid.
</para>
<para>
<command>const string getMonthName()</command>
returns month as a string in the format of January...December
</para>
<para>
<command>const string getShortMonthName()</command>
returns month as a string in the format of Jan...Dec
</para>
<para>
<command>const string getYear()</command>
returns year in the format of YYYY. A return value of -1 indicates
an error because the date entered was invalid.
</para>
<para>
<command>const string getShortDate()</command>
returns date in short date in the format of DD/MM/YY
</para>
<para>
<command>const string getLongDate()</command>
returns date in long date date in the format of DD MMMMMM YY
</para>
<para>
<command>const string getDecimalDate()</command>
returns the date as a decimal date in the form YYYY.DDDD. A return value of -1 indicates
an error because the date entered was invalid.
</para>
<para>
2 new methods added yearsToJulian and julianToYears. Both take and return doubles
and have no relation to the member variables of this object.
</para>
.SH "VERSION"
.PP
This documentation was generated for the \fBGeodetic Data Manipulation System\fR 0.1 by \fBautodocbook\fR (http://www.stillhq.com).
.SH "AUTHOR"
.PP
The \fBGeodetic Data Manipulation System\fR is under development by Michael Still (mikal@stillhq.com), Daniel Fernandez, Kristy Van Der Vlist, Blake Swadling, and Nick Wheatstone. All code is Copyright 2002 Michael Still, Daniel Fernandez, Kristy Van Der Vlist, Blake Swadling and Nick Wheatstone,  and is released under the GNU GPL. Code submissions are welcomed. Contact Michael at the address above if you have something you would like to contribute.
.SH "BUGS"
.PP
There  are no known bugs in the \fBGeodetic Data Manipulation System\fR. If you find one, please contact mikal@stillhq.com and let me know.


cep/ChangeLog
2002-02-27 Michael Still mikal@stillhq.com
	Diarised the meeting of 26 February 2002

2002-03-01 Michael Still mikal@stillhq.com
	Diarised the meeting from today

2002-03-01 Michael Still mikal@stillhq.com
	Started work on the intial report for the project proposal... This was mainly building the skeleton of the document which will be fleshed out at the weekend meeting

2002-04-29 Daniel
	Added cepWindows class
	
2002-05-01 Michael Still mikal@stillhq.com
	Meeting reports

2002-05-01 Michael Still mikal@stillhq.com
	Added cepWindows to build process and fixed compile errors

2002-05-04 Daniel Fernandez daniel.fernandez@ise.canberra.edu.au
	Fixed some todos in cepWindows.cpp and began implementation of 
	vector for using data and arrow of windows.

2002-05-04 Blake Swadling blake@swadling.org
	Added the source files and ps files for the initial propject proposal.
	Added matlab files for preliminary prototype

2002-05-06 Daniel Fernandez daniel.fernandez@ise.canberra.edu.au
	Further addressed some todos.
	changed implementation of vectors to suit definition. Not yet 
	complete.

2002-05-07 Daniel Fernandez daniel.fernandez@ise.canberra.edu.au
	Addressed more todos.
	Fixed implementation of vectors in both windows and ham.

2002-05-07 Daniel Fernandez daniel.fernandez@ise.canberra.edu.au
	Added constructor to cepDataset.cpp to deal with window and 
	hamming return paramaters.

2002-05-13 Daniel Fernandez daniel.fernandez@ise.canberra.edu.au
	Added resize functionality to vector in cepWindows.cpp
	Addressed more todos in order add cepWindows.cpp code to cepDataset.cpp
	Added cepWindows.cpp code to cepDataset.cpp	

2002-05-13 Daniel Fernandez daniel.fernandez@ise.canberra.edu.au
	Added the fftw package (tarballed) in cvs001/fftw/fftw-2.1.3
	Added some general fftw papers in cvs001/fftw

2002-05-18 Daniel Fernandez daniel.fernandez@ise.canberra.edu.au
	Changed the vector data sturcture through out cepDataset
	some general fixes necessary...
	With Michael and Kristy.

2002-05-20 Daniel Fernandez daniel.fernandez@ise.canberra.edu.au
	continued changing/fixes for the new vector data sturcture 
	through out cepDataset and some general fixes...

2002-05-20 Daniel Fernandez daniel.fernandez@ise.canberra.edu.au
	added meeting minutes for weeks 13, 14, 15. Still more to come.

2002-06-13 Daniel Fernandez daniel.fernandez@ise.canberra.edu.au
	modified week13 minutes to include meeting with Peter Morgan
	modified configure.in to to link fftw libraries
	committed configure after re running autoconf
	
2002-07-23 Kristy Van Der Vlist bobhchicken@bigpond.com
	added buildreport.pl to the meetings directory. This script should be
	used to generate the meeting reports rather than buildreport.sh as it
	fixes the problems associated with the first one. 

2002-08-04 Kristy Van Der Vlist bobhchicken@bigpond.comi
	added final version of cepMatrix to source. Also added updated version
	of cepLs - the residual calcs are to come.

2002-08-16 Daniel Fernandez daniel.fernandez@ise.canberra.edu.au
	added fft template class and test code



cep/source/cepDate.cpp
/* 
   Imp for the CEP date
   Copyright (C) Michael Still                    2002
   
   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.
   
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
   
   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#include "core.h"

const string cMonthNames[12] =
  { "January", "February", "March", "April", "May",
  "June", "July", "August", "September", "October",
  "November", "December"
};

cepDate::cepDate (double decimal):
m_day (-1),
m_year (-1),
m_yearDays (-1),
m_dayOfYear (-1),
m_month (-1),
m_leap (false)
{
  m_decimal = decimal;
}

void
cepDate::popDayOfYear ()
{
  popYear ();
  popDaysInYear ();

  if (m_dayOfYear == -1)
    {
      m_dayOfYear = m_yearDays * (m_decimal - m_year);
      cepDebugPrint (cepDtoa (m_decimal) + " converts to " + cepItoa (m_dayOfYear) + " days");
    }
}

void
cepDate::popDaysInYear ()
{
  popYear ();

  if (m_yearDays == -1)
    {
      // This is Blake's fault -- he made me do it
      // todo_mikal: Change days in years
      m_yearDays = ((m_year % 100 == 0) && (m_year % 400 == 0)) ||
	((m_year % 100 != 0) && (m_year % 4 == 0)) ? 366 : 365;

      cepDebugPrint (cepDtoa (m_decimal) + " converts to " + 
		     cepItoa (m_yearDays) + " days in the year");

      if (m_yearDays == 366)
	m_leap = true;
    }
}

void
cepDate::popYear ()
{
  if (m_year == -1)
    {
      m_year = (int) m_decimal;

      cepDebugPrint (cepDtoa (m_decimal) + " converts to " + cepItoa (m_year) +
		     " year");
    }
}

void
cepDate::popMonthAndDay ()
{
  popDayOfYear ();
  popYear ();
  popDaysInYear ();

  if (m_day == -1)
    {
      int dayOfYear = m_dayOfYear;
      int i = 0;

      int months[12] = { 31, m_leap ? 29 : 28, 31, 31, 31, 31, 31, 31, 31, 31,
	31, 31
      };

      while ((i < 11) && (dayOfYear - months[i] > 0))
	{
	  dayOfYear -= months[i];
	  i++;
	}
      m_day = dayOfYear;
      m_month = i;

      cepDebugPrint (cepDtoa (m_decimal) + " converts to " + cepItoa (m_day) +
		     " " + cMonthNames[m_month]);
    }
}

string cepDate::toString ()
{
  popMonthAndDay ();
  popYear ();

  return string (cepItoa (m_day) + " " + cMonthNames[m_month] + " " +
		 cepItoa (m_year));
}
